programs/wen_transfer_guard/README.md

```diff
@@ -0,0 +1,147 @@
+# Wen Transfer Guard Program
```

Comment:
### Wen Transfer Guard Program

The Wen Transfer Guard Program secures token transfers on the Solana blockchain by enforcing customizable rules using the token_2022 program transfer hook interface.

## How It Works

### Overview

1. **Create a Transfer Guard Account**: 
   Set up an account with a ruleset and identifiable metadata.

2. **Update the Guard**: 
   Modify the ruleset of the guard account.

3. **Initialize the Guard**: 
   Link the guard to a specific mint.

4. **Execute Transfer Rules**: 
   Enforce the ruleset during token transfers.

### Instructions Summary

- **create_guard**: Creates a new transfer guard account.
- **update_guard**: Updates the ruleset of a transfer guard account.
- **initialize**: Assigns extra metas to a given mint, linking the guard to it.
- **execute**: Enforces the ruleset during a token transfer.

### Guard Ruleset Modes

- **Metadata Rules**: Validate against custom metadata fields.
- **Amount Rules**: Enforce limits on the amount of tokens being transferred.
- **CPI Rules**: Control which programs can interact with the mint’s tokens.

### Example Flow (Anchor Based)

#### Creating a Guard

```ts
const guardMint = web3.Keypair.generate();

const guardMintAta = getAssociatedTokenAddressSync(
  guardMint.publicKey,
  guardAuthority,
  false,
  TOKEN_2022_PROGRAM_ID
);

const ix = await program.methods
  .createGuard({
    name: "Guard",
    symbol: "GRD",
    uri: "https://example.com/metadata.json",
    additionalFieldsRule: [],
    transferAmountRule: null,
    cpiRule: {
      deny: { 0: [new web3.PublicKey("11111111111111111111111111111111")] },
    },
  })
  .accounts({
    mint: guardMint.publicKey,
    mintTokenAccount: guardMintAta,
    guardAuthority,
    payer,
  })
  .instruction();

const txId = await sendSignedVtx(
  provider,
  payer.publicKey,
  [payer, guardAuthority, guardMint],
  ix
);
```

#### Updating a Guard

```ts
const ix = await program.methods
  .updateGuard({
    additionFieldsRule: [],
    transferAmountRule: null,
    cpiRule: {
      deny: { 0: [] },
    },
  })
  .accounts({
    mint: guardMint.publicKey,
    tokenAccount: guardMintAta,
    guardAuthority: kGuardOwner.publicKey,
  })
  .instruction();

const txId = await sendSignedVtx(provider, payer, [guardAuthority], ix);
```

#### Initializing (Assign Guard to Mint)

```ts
const [guardAddress] = web3.PublicKey.findProgramAddressSync(
  [
    Buffer.from("wen_token_guard"),
    Buffer.from("guard_v1"),
    guardMint.publicKey.toBuffer(),
  ],
  program.programId
);
const extraMetasAddress = getExtraAccountMetaAddress(
  mint.publicKey,
  program.programId
);

const ix = await program.methods
  .initialize()
  .accountsStrict({
    guard: guardAddress,
    mint: mint.publicKey,
    mintAuthority: mintAuthority.publicKey,
    payer: payer.publicKey,
    extraMetasAccount: extraMetasAddress,
    systemProgram: web3.SystemProgram.programId,
  })
  .instruction();

await sendSignedVtx(provider, payer.publicKey, [payer, mintAuthority], ix);
```

#### Executing a Transfer

```ts
let ix = await createTransferCheckedWithTransferHookInstruction(
  provider.connection,
  sourceAta,
  mint.publicKey,
  destinationAta,
  sourceAuthority.publicKey,
  BigInt(1e8),
  mint.decimals,
  undefined,
  undefined,
  TOKEN_2022_PROGRAM_ID
);

await sendSignedVtx(
  provider,
  context.payer.publicKey,
  [kSourceAuthority, context.payer],
  ix
);
```

Reply:
An updated README to swap out @kevinrodriguez-io